framework/model/source/k8s/metric_source.go

```diff
@@ -0,0 +1,306 @@
+	prometheus "github.com/prometheus/client_golang/api/prometheus/v1"
+	prometheusModel "github.com/prometheus/common/model"
+	log "github.com/sirupsen/logrus"
+	"istio.io/api/networking/v1alpha3"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"reflect"
+	bootconfig "slime.io/slime/framework/apis/config/v1alpha1"
+	"slime.io/slime/framework/bootstrap"
+	"slime.io/slime/framework/controllers"
+	"slime.io/slime/framework/model"
+	"slime.io/slime/framework/model/module"
+	"slime.io/slime/framework/util"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type MetricSource struct {
+	name                     string
+	gvks                     []schema.GroupVersionKind
+	metricUpdateCheckHandler func(event model.WatcherEvent) map[string]*bootconfig.Prometheus_Source_Handler
+	moduleEventChan          chan<- model.ModuleEvent
+	restConfig               rest.Config
+	env                      *bootstrap.Environment
+	api                      prometheus.API
+	interest                 cmap.ConcurrentMap
+	sync.RWMutex
+}
+
+// Register Module -> MetricSource
+func Register(mo module.Module) (*MetricSource, error) {
+	log := log.WithField("Module Register", mo.Name())
+	m := &MetricSource{}
+	m.name = mo.Name()
+	m.gvks = mo.GVKs()
+	m.metricUpdateCheckHandler = mo.MetricUpdateCheckHandler()
+	m.moduleEventChan = mo.ModuleEventChan()
+	m.restConfig = mo.RestConfig()
+	m.env = mo.Env()
+
+	// init m.api from env
+	if ps := mo.Env().Config.Metric.Prometheus; ps != nil {
+		promClient, err := prometheusApi.NewClient(prometheusApi.Config{
+			Address:      ps.Address,
+			RoundTripper: nil,
+		})
+		if err != nil {
+			log.Errorf("failed to create prometheus client, %+v", err)
+		} else {
+			m.api = prometheus.NewAPI(promClient)
+			log.Infof("successfully create prometheus client")
+		}
+	}
+
+	// init m.interest
+	m.interest = cmap.New()
+
+	return m, nil
+}
+
+func (m *MetricSource) getInterest() cmap.ConcurrentMap {
+	return m.interest
+}
+
+// InterestAdd add resource to interest
+func (m *MetricSource) InterestAdd(nn types.NamespacedName) {
+	m.interest.Set(nn.Namespace+"/"+nn.Name, true)
+}
+
+// InterestRemove remove from interest, the resource will be deleted by k8s
+func (m *MetricSource) InterestRemove(nn types.NamespacedName) {
+	m.interest.Pop(nn.Namespace + "/" + nn.Name)
+}
+
+func (m *MetricSource) interestGet(nn types.NamespacedName) bool {
+	_, result := m.interest.Get(nn.Namespace + "/" + nn.Name)
+	return result
+}
+
+func (m *MetricSource) Name() string {
+	return m.name
+}
+
+func (m *MetricSource) RestConfig() rest.Config {
+	return m.restConfig
+}
+
+func (m *MetricSource) GVKs() []schema.GroupVersionKind {
+	return m.gvks
+}
+
+// Notify push ModuleEvent to module
+func (m *MetricSource) Notify(we model.WatcherEvent) {
+	// event preCheck
+	if m.gvkCheck(we.GVK) && m.interestGet(we.NN) {
```

Comment:
early return is a more recommended way in golang.

if !m.gckCheck(we.GVK) || !m.interestGet(we.NN) {
  return
}